### PR TITLE
Feature/jwt auth

### PR DESCRIPTION
**Pull request by longngoryu**

### CODE REVIEW

File: server/src/app/auth/routes.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import multer from "multer";
+import * as controllers from "./controllers";
+
+const upload = multer();
+const router = Router();
+
+router.post("/signup", upload.none(), controllers.signup);
+router.post("/signin", upload.none(), controllers.signin);
+router.post("/refresh", upload.none(), controllers.refresh);
+
```
[the remark this answers is not in the document]
Comment (longngoryu) [reply server/src/app/auth/routes.ts:11]: yes sir

File: server/src/app/auth/controllers.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+  if (!email || !password) {
+    return res.status(409).json({ error: "Email and password not empty!" });
+  }
+
+  try {
+    const saltRounds = 10;
+    const hash = bcrypt.hashSync(password, saltRounds);
+
+    const user = await User.create({
```
[the remark this answers is not in the document]
Comment (longngoryu) [reply server/src/app/auth/controllers.ts:18]: Can I create it in the src/config/constant.ts folder?

File: server/src/app/auth/controllers.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+  if (!user) {
+    return res.status(404).json({ error: "Not found user!" });
+  }
+
+  const isCompare = bcrypt.compareSync(password, user.dataValues.password);
+
+  if (!isCompare) {
+    return res.status(401).json({ error: "Login failed!" });
+  }
```
[the remark this answers is not in the document]
Comment (longngoryu) [reply server/src/app/auth/controllers.ts:46]: yes sir

File: server/src/app/auth/controllers.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    const user = await User.create({
+      name: name,
+      email: email,
+      password: hash,
+    });
+
+    return res.status(201).json(user);
+  } catch (error) {
+    return res.status(400).json(error);
+  }
```
[the remark this answers is not in the document]
Comment (longngoryu) [reply server/src/app/auth/controllers.ts:27]: Which number should I use?

File: server/src/middleware/token.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { Request, Response, NextFunction } from "express";
+import jwt from "jsonwebtoken";
+import env from "../config/env";
+
+export async function verify(req: Request, res: Response, next: NextFunction) {
```
[the remark this answers is not in the document]
Comment (longngoryu) [reply server/src/middleware/token.ts:5]: yes sir

File: server/src/app/users/routes.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import { Router } from "express";
+import { getUsers } from "./controllers";
+import { verify } from "../../middleware/token";
+
+const router = Router();
+
+router.get("/", verify, getUsers);
+
```
[the remark this answers is not in the document]
Comment (longngoryu) [reply server/src/app/users/routes.ts:8]: yes

File: server/src/app/auth/controllers.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  return res.status(200).json({
+    access: tokenAccess,
+    refresh: tokenRefresh,
+  });
+}
+
+export async function refresh(req: Request, res: Response) {
+  if (!req.headers.authorization) {
+    return res.status(403).json({ error: "Token not found!" });
+  }
```
[the remark this answers is not in the document]
Comment (longngoryu) [reply server/src/app/auth/controllers.ts:70]: Should I change the message?
 "Token is required!"

File: server/src/app/auth/controllers.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+import jwt from "jsonwebtoken";
+import env from "../../config/env";
+import User from "../users/models";
+
+export async function signup(req: Request, res: Response) {
+  const { name, email, password } = req.body;
+
+  if (!email || !password) {
+    return res.status(409).json({ error: "Email and password not empty!" });
+  }
```
[the remark this answers is not in the document]
Comment (longngoryu) [reply server/src/app/auth/controllers.ts:12]: yes sir

File: server/src/app/auth/routes.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (longngoryu) [reply server/src/app/auth/routes.ts:11]: yes

File: server/src/app/auth/controllers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (minhnhat-dev) [reply server/src/app/auth/controllers.ts:18]: You can create into app/auth/auth.constant.ts

File: server/src/app/auth/controllers.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (minhnhat-dev) [reply server/src/app/auth/controllers.ts:27]: I think code 500[link](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/500)

File: server/src/app/auth/controllers.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (minhnhat-dev) [reply server/src/app/auth/controllers.ts:70]: I think message just "Unauthorized"

**Comment by minhnhat-dev** [comment]
@longngoryu  I've given you comments, Please resolve it!

**Pull request merged**